naga/src/proc/constant_evaluator.rs

```diff
@@ -2497,6 +2520,121 @@ impl<'a> ConstantEvaluator<'a> {
 
         Ok(resolution)
     }
+
+    fn select(
+        &mut self,
+        reject: Handle<Expression>,
+        accept: Handle<Expression>,
+        condition: Handle<Expression>,
+        span: Span,
+    ) -> Result<Handle<Expression>, ConstantEvaluatorError> {
+        let arg = |this: &mut Self, expr| {
+            let expr = this.check_and_get(expr)?;
+            let expr = this.eval_zero_value_and_splat(expr, span)?;
+            let expr: &Expression = &this.expressions[expr];
+            this.try_eval_and_append(expr.clone(), span)
+        };
+
+        let reject = arg(self, reject)?;
+        let accept = arg(self, accept)?;
+        let condition = arg(self, condition)?;
+
+        let select_single_component =
+            |this: &mut Self, reject_scalar, reject, accept, condition| {
+                let accept = this.cast(accept, reject_scalar, span)?;
+                if condition {
+                    Ok(accept)
+                } else {
+                    Ok(reject)
+                }
+            };
+
+        match (&self.expressions[reject], &self.expressions[accept]) {
```

Comment:
In the case where the condition is `bool` and not `vecN<bool>`, can't we just return `reject` or `accept` directly, without ever looking at its value?

Reply:
Is it not possible to reuse any of the `component_wise` machinery for this?

Reply:
It wasn't possible previous because we're using a `cast` operation before performing a component-wise operation here; we already take on the scope of delving into all components before knowing we have the right types on LHS and RHS.

I might have something wrong here, though.

Reply:
At least with my current arguments, this simplification depends on whether we can remove `cast` at this conversation: https://github.com/gfx-rs/wgpu/pull/7602#discussion_r2067328197